pymc_marketing/mmm/hsgp.py

```diff
@@ -0,0 +1,967 @@
+    lengthscale_range: tuple[float, float],
+    cov_func: str,
+) -> tuple[int, float]:
+    """Use heuristics for minimum `m` and `c` values.
+
+    Based on recommendations from Ruitort-Mayol et. al.
+
+    In practice, you need to choose `c` large enough to handle the largest
+    lengthscales, and `m` large enough to accommodate the smallest lengthscales.
+
+    NOTE: These recommendations are based on a one-dimensional GP.
+
+    Parameters
+    ----------
+    x : tensor_like
+        The x values the HSGP will be evaluated over.
+    x_center : tensor_like
+        The center of the data.
+    lengthscale_range : tuple[float, float]
+        The range of the lengthscales. Should be a list with two elements [lengthscale_min, lengthscale_max].
+    cov_func : str
+        The covariance function to use. Supported options are "expquad", "matern52", and "matern32".
+
+    Returns
+    -------
+    - `m` : int
+        Number of basis vectors. Increasing it helps approximate smaller lengthscales, but increases computational cost.
+    - `c` : float
+        Scaling factor such that L = c * S, where L is the boundary of the approximation.
+        Increasing it helps approximate larger lengthscales, but may require increasing m.
+
+    Raises
+    ------
+    ValueError
+        If either `x_range` or `lengthscale_range` is not in the correct order.
+
+    References
+    ----------
+    .. [1] Ruitort-Mayol, G., Anderson, M., Solin, A., Vehtari, A. (2022).
+    Practical Hilbert Space Approximate Bayesian Gaussian Processes for Probabilistic Programming
+    """
+    lengthscale_min, lengthscale_max = lengthscale_range
+    if lengthscale_min >= lengthscale_max:
+        raise ValueError(
+            "The boundaries are out of order. {lengthscale_min} should be less than {lengthscale_max}"
+        )
+
+    Xs = x - x_center
+    S = np.max(np.abs(Xs), axis=0)
+
+    if cov_func.lower() == "expquad":
+        a1, a2 = 3.2, 1.75
+
+    elif cov_func.lower() == "matern52":
+        a1, a2 = 4.1, 2.65
+
+    elif cov_func.lower() == "matern32":
+        a1, a2 = 4.5, 3.42
+
+    else:
+        raise ValueError(
+            "Unsupported covariance function. Supported options are 'expquad', 'matern52', and 'matern32'."
+        )
+
+    c = max(a1 * (lengthscale_max / S), 1.2)
+    m = int(a2 * c / (lengthscale_min / S))
+
+    return m, c
+
+
+class CovFunc(str, Enum):
+    """Supported covariance functions for the HSGP model."""
+
+    ExpQuad = "expquad"
+    Matern52 = "matern52"
+    Matern32 = "matern32"
+
+
+def create_eta_prior(mass: float = 0.05, upper: float = 1.0) -> Prior:
+    """Create prior for the variance.
+
+    Parameters
+    ----------
+    mass : float
+        Mass of the variance. Default is 0.05 or 5%.
+    upper : float
+        Upper bound for the variance. Default is 1.0.
+
+    """
+    return Prior(
+        "Exponential",
+        lam=-pt.log(mass) / upper,
+    )
+
+
+def create_m_and_L_recommendations(
```

Comment:
Can't we use https://github.com/pymc-devs/pymc/blob/main/pymc/gp/hsgp_approx.py#L97 ?

Reply:
Can we add some references on where these heuristic come from (just the paper reference, ideally the section)

Reply:
(I think we can use the same ref as above)